packages/composite-checkout-wpcom/src/components/checkout.jsx

```diff
@@ -23,13 +23,11 @@ const failureRedirectUrl = window.location.href;
  * Constructs a cart endpoint callback
  *
  * @param {object} wpcom Not sure what this is.
- * @returns {Function} Server callback
+ * @returns {Function} Server callback, returning a promise
  */
-export function callWpcomShoppingCartEndpoint( wpcom ) {
-	return async function( cartKey, cartParam ) {
-		return await wpcom.req.get( `/me/shopping-cart/${ cartKey }`, cartParam );
-	};
-}
+export const callWpcomShoppingCartEndpointWith = wpcom => async ( cartKey, cartParam ) => {
+	return await wpcom.setCart( cartKey, cartParam );
```

Comment:
This function is `async`, which means it will always return a Promise. Inside it we are calling `wpcom.setCart()`, which also returns a Promise. When it's called, we're using `await` on that promise, which will wait for that Promise to complete before continuing. Once that Promise resolves, the outer function will return the resulting value in its own Promise, but that Promise will be resolved immediately.

If this were written without async/await, it would look like this:

```js
export const callWpcomShoppingCartEndpointWith = wpcom => ( cartKey, cartParam ) => {
  return new Promise( ( resolve, reject ) => {
    wpcom.setCart( cartKey, cartParam )
      .then( result => resolve( result ) )
      .catch( error => reject( error ) );
  } );
};
```

It's unnecessary to wait for the inner Promise to resolve before returning the outer Promise since there's no other execution that needs to occur after the inner Promise resolves. Instead of creating a second Promise, we could just return the inner Promise, like so:

```js
export const callWpcomShoppingCartEndpointWith = wpcom => ( cartKey, cartParam ) => {
  return wpcom.setCart( cartKey, cartParam );
};
```

Or, even as a one-liner:

```js
export const callWpcomShoppingCartEndpointWith = wpcom => ( cartKey, cartParam ) => 
  wpcom.setCart( cartKey, cartParam );
```

If you wanted to, you could even remove the forwarding function and just return the inner function since the arguments are the same:

```js
export const callWpcomShoppingCartEndpointWith = wpcom => wpcom.setCart;
```

Reply:
Thanks for the thorough explanation- that helps a lot. I was kind of flailing around putting async/await in different places trying to figure out where the server response was going, but it looks like your fix in ea9f8f5 addresses the real problem.

Reply:
Also- I like this a lot:

```
export const callWpcomShoppingCartEndpointWith = wpcom => wpcom.setCart;
```

However I'd also really like for the signature of this function to be manifest without needing to chase down where `wpcom.setCart` comes from. But while thinking about how to do that I realized that this function is only used in one place, and there's no need to pass the entire `wpcom` object in to this code- the whole point of injecting `wpcom.setCart` is to abstract out the call, and the `wpcom` object comes with... a lot of other stuff too. So I instead inlined it at the only call site like this:

```
const useShoppingCart = makeShoppingCartHook(
    ( cartKey, cartParam ) => wpcom.setCart( cartKey, cartParam ),
    initialCart
);
```

(It's probably possible to write that without the explicit arguments, but personally I prefer it that way in the absence of a static type signature)

Reply:
I think you're absolutely right. It's a paint to have to search out the function definition to find out what the arguments are. 👍